types/aws-lambda/trigger/api-gateway-proxy.d.ts

```diff
@@ -66,6 +93,60 @@ export interface APIGatewayProxyResult {
     isBase64Encoded?: boolean;
 }
 
+/**
+ * Works with HTTP API integration Payload Format version 2.0
+ * @see - https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html
+ */
+export interface APIGatewayProxyEventV2 {
+    routeKey: string;
+    rawPath: string;
+    rawQueryString: string;
+    cookies: string[];
+    headers: { [name: string]: string };
+    queryStringParameters: { [name: string]: string };
+    requestContext: {
+      accountId: string;
+      apiId: string;
+      authorizer: { jwt: {
+          claims: { [name: string]: string };
+          scopes: string[]
+          }
+      };
+      domainName: string;
+      domainPrefix: string;
+      http: {
+        method: string;
+        path: string;
+        protocol: string;
+        sourceIp: string;
+        userAgent: string;
+      };
+      requestId: string;
+      routeKey: string;
+      stage: string;
+      time: string;
+      timeEpoch: number;
+    };
+    body: string;
+    pathParameters: { [name: string]: string };
+    isBase64Encoded: boolean;
+    stageVariables: { [name: string]: string };
+    }
+
+/**
+ * Works with HTTP API integration Payload Format version 2.0
+ * @see - https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html
+ */
+export interface APIGatewayProxyResultV2 {
+    statusCode?: number;
+    headers?: {
+        [header: string]: boolean | number | string;
+    };
+    body?: string | object;
+    isBase64Encoded?: boolean;
+    cookies?: string[];
+}
```

Comment:
AWS document that you can also return the body to be JSONed, or the JSON string to use as the body, but not sure it's worth the loss of typing? Possibly this is `APIGatewayProxyV2StrictEvent`, and `type APIGatewayProxyV2Event = unknown | string | APIGatewayProxyV2StrictEvent`, which I believe will still give completion, but no type checking.

Reply:
Unlike v1, v2 is smart enough to automatically detect JSON or text and add appropriate `Content-Type` header. No longer need to stringify a JSON object.

Reply:
You may be misreading the doc page? There's at least no example of returning an object in `body` property, only of directly returning the object, returning the string (which is assumed to be JSON) or returning this type.

Reply:
I am not sure if I understood you, but I've tested it myself. Here are the results:

- If I assign string to `body`, it infers `statusCode` as `200` and sets the `Content-Type` to `text/plain; charset=utf-8` even if the body content was output of `JSON.stringify()`.

- If I assign an object to `body` (without doing `JSON.stringify()`), it infers `statusCode` as `200` and sets `Content-Type` to `application/json`.

- If I assign boolean, number, null, undefined to `body` without `statusCode`, it misbehaves by returning the whole result object with `Content-Type` set to `application/json`.

- If I assign boolean, number, null, undefined to `body` with `statusCode`, content is returned as text with `Content-Type` set to `text/plain; charset=utf-8`.

Reply:
Thanks for testing that, and it's nice to know, but the documentation claims in the "Lambda function response for format 2.0" section that you can return a JSON string or object directly, instead of as a body property, and it will interpret it as the `body` property of a JSON response, stringifying it if it is not a string - that's what I was referring to.

Since you can directly return the body value now, unrelated to the above issue, I'd like a way to type the `body` property e.g. `APIGatewayProxyResultV2<MyHttpJsonBody>`, but it's hard to cleanly make `body` required only if you provide a type parameter. Take a stab if you like, but not a priority.